src/parser/SpatialQuery.cpp

```diff
@@ -0,0 +1,149 @@
+SpatialJoinConfiguration SpatialQuery::toSpatialJoinConfiguration() const {
+  if (!left_.has_value()) {
+    throw SpatialSearchException("Missing parameter 'left' in spatial search.");
+  } else if (!maxDist_.has_value() && !maxResults_.has_value()) {
+    throw SpatialSearchException(
+        "Neither 'nearestNeighbors' nor 'maxDistance' were provided. At least "
+        "one of both is required.");
+  } else if (!right_.has_value()) {
+    throw SpatialSearchException(
+        "Missing parameter 'right' in spatial search.");
+  }
+
+  // Only if the number of results is limited, it is mandatory that the right
+  // variable must be selected inside the service. If only the distance is
+  // limited, it may be declared inside or outside of the service.
+  if (maxResults_.has_value() && childGraphPattern_._graphPatterns.empty()) {
+    throw SpatialSearchException(
+        "Missing parameter 'right' in spatial search. A spatial search with "
+        "a maximum number of results must have its right variable declared "
+        "inside the service using a graph pattern: SERVICE spatialSearch: { "
+        "... { ... ?right } }.");
+  }
+
+  // Default algorithm
+  SpatialJoinAlgorithm algo = SPATIAL_JOIN_DEFAULT_ALGORITHM;
+  if (algo_.has_value()) {
+    algo = algo_.value();
+  }
+
+  if (maxResults_.has_value()) {
+    return SpatialJoinConfiguration{
+        NearestNeighborsConfig{maxResults_.value(), maxDist_}, left_.value(),
+        right_.value(), bindDist_, algo};
+  } else {
+    return SpatialJoinConfiguration{MaxDistanceConfig{maxDist_.value()},
+                                    left_.value(), right_.value(), bindDist_,
+                                    algo};
+  }
+}
+
+// ____________________________________________________________________________
+SpatialQuery::SpatialQuery(const SparqlTriple& triple) {
+  AD_CONTRACT_CHECK(triple.p_.isIri(),
+                    "The config triple for SpatialJoin must have a special IRI "
+                    "as predicate.");
+  const std::string& input = triple.p_._iri;
+
+  if (input.starts_with(NEAREST_NEIGHBORS)) {
+    throw SpatialSearchException(
+        "The special predicate <nearest-neighbors:...> is no longer supported "
+        "due to confusing semantics. Please use SERVICE spatialSearch: {...} "
+        "instead. For information on its usage, see the QLever Wiki.");
+  }
+
+  // Add variables to configuration object
+  AD_CONTRACT_CHECK(triple.s_.isVariable() && triple.o_.isVariable(),
+                    "Currently, SpatialJoin needs two variables");
+  setVariable("left", triple.s_, left_);
+  setVariable("right", triple.o_, right_);
+
+  // Helper to convert a ctre match to an integer
+  auto matchToInt = [](std::string_view match) -> std::optional<size_t> {
+    if (match.size() > 0) {
+      size_t res = 0;
+      std::from_chars(match.data(), match.data() + match.size(), res);
+      return res;
+    }
+    return std::nullopt;
```

Comment:
I think this can never happen, because the regex expects a number with at least one digit.
So you can AD_CORRECTNESS_CHECK(match.size() > 0) in the beginning of the lambda, and everything stays the same.